### ISSUE TITLE
Non-Affine Transformation not working with multmatrix()

### ISSUE THREAD

**Issue by t-paul** [openscad/openscad#749]
Report from IRC:

I'm not sure if non-affine transformation is supported but multmatrix() seems to indicate it should work. The following example does not work as expected:

```
    // trying to achieve non-affine transformations fails
    // http://www.charlespetzold.com/blog/2007/08/060605.html

    object();

    module object()
    {
            //cylinder(20,10,10);
            //translate([0,0,20]) cube([5,15,5]);
            cube([15,15,15]);
    }

    translate([25,0,0]) multmatrix(m=[
            [1,0,0,0],
            [0,10,0,9],
            [0,0,1,0],
            [0,0,0,1]]
            ) object();
```
##

**Comment by ioquatix** [comment]
I was trying to do non-affine transform and it wasn't obvious that it wouldn't work until I finally wondered if it wasn't implemented and found this. If the matrix is non-affine a warning might be a good idea if this is not going to be implemented.

**Comment by GilesBathgate** [comment]
Like @NateTG says this should be possible to code. Just get the points apply the transformation, you would first have to get the points from the CGAL NefPolyhedron, apply the transform matrix (perhaps with Eigen, although it needs to be with exact number type), and then either create a new Nef with these transformed points or move the existing ones. To move the exsiting ones, you could use the approach I use here: https://github.com/GilesBathgate/RapCAD/blob/master/src/cgalprimitive.cpp#L552

**Comment by doug-moen** [comment]
I wonder why CGAL doesn't support non-affine transformation matrices in the first place? Must be a reason.

It occurs to me that a non-affine transformation can map several distinct points onto the same point. So if you do this to a NefPolyhedron, then does that violate a data structure invariant? Could it lead to a crash? Personally I dunno.

**Comment by GilesBathgate** [comment]
@doug-moen Yes, it could, so you would also probably need to do some cleanup. FYI CGAL::Aff_transformation_3 just dosen't support "the bottom row": http://doc.cgal.org/latest/Kernel_23/classCGAL_1_1Aff__transformation__3.html#aa9c9cce6a5b44397a1af5351cebfaaaa

**Comment by doug-moen** [comment]
AFAIK, a perspective transformation matrix could cause damage if the "eye point" or "vanishing point" is inside the object, or on the surface.

There are two kinds of damage that this could cause.
 1. It could cause a face to become degenerate (zero area). Fixing this could involve merging two co-incident vertices, or splitting a face where a vertex has landed in the middle of an edge.
 2. It could introduce self intersections. Fixing this involves adding vertices and splitting faces.

Fixing the mesh is easier if all the faces are triangles: less case analysis. So here's my algorithm:
```
if (the transformation matrix is non-affine) {
    triangulate the mesh;
    transform each vertex;
    repair the mesh;
}
```
Hopefully we can just steal an open source mesh repair algorithm from somewhere.

**Comment by kintel** [comment]
@doug-moen There's a brl-cad mesh healing project going on as part of GSoC: http://brlcad.org/wiki/User:Tandoorichick/GSoC2016/Project

The developer is targeting OpenSCAD as well. IRC is where this stuff is happening.
